docs/map.jinja.rst

```diff
@@ -0,0 +1,498 @@
+
+          # default values
+          - "Y:G@id"
+    ...
+
+The syntax is explained later at `Sources of configuration values`_.
+
+
+Bind roles to minions
+`````````````````````
+
+We associate roles `grains`_ to minion using `grains.append`_.
+
+For the servers:
+
+.. code-block:: console
+
+    salt 'server-*' grains.append roles template/server
+
+For the clients:
+
+.. code-block:: console
+
+    salt 'client-*' grains.append roles template/client
+
+.. note::
+
+    Since we used ``Y:C@roles``, ``map.jinja`` will do a ``salt['config.get']('roles')`` to retrieve the roles so you could use any other method to bind roles to minions (`pillars`_ or `SDB`_) but `grains`_ seems to be the prefered method.
+
+Note for Microsoft Windows systems
+``````````````````````````````````
+
+If you have a minion running under windows, you can't use colon ``:`` as a delimiter for grain path query (see `bug 58726`_) in which case you should use an alternate delimiter:
+
+Modify ``/srv/salt/parameters/map_jinja.yaml`` to change the query for ``dns:domain`` to define the `alternate delimiter`_:
+
+.. code-block:: yaml
+
+    ---
+    values:
+      map_jinja:
+        sources:
+          # default values
+          - "Y:G@osarch"
+          - "Y:G@os_family"
+          - "Y:G@os"
+          - "Y:G@osfinger"
+          - "C@{{ tplroot ~ ':lookup' }}"
+          - "C@{{ tplroot }}"
+
+          # Roles activate/deactivate things
+          # then thing are configured depending on environment
+          # So roles comes before `dns:domain`, `domain` and `id`
+          - "Y:C@roles"
+
+          # DNS domain configured (DHCP or resolv.conf)
+          - "Y:G:!@dns!domain"
+
+          # Based on minion ID
+          - "Y:G@domain"
+
+          # default values
+          - "Y:G@id"
+    ...
+
+And then, rename the directory:
+
+.. code-block:: console
+
+    mv /srv/salt/TEMPLATE/parameters/dns:domain/  '/srv/salt/TEMPLATE/parameters/dns!domain/'
+
+
+Format of configuration YAML files
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When you write a new YAML file, note that it must conform to the following layout:
+
+- a mandatory ``values`` key to store the configuration values
+- two optional keys to configure the use of `salt.slsutil.merge`_
+
+  - an optional ``strategy`` key to configure the merging strategy, for example ``strategy: 'recurse'``, the default is ``smart``
+  - an optional ``merge_lists`` key to configure if lists should be merged or overridden for the ``recurse`` and ``overwrite`` strategy, for example ``merge_lists: 'true'``
+
+Here is a valid example:
+
+.. code-block:: yaml
+
+    ---
+    strategy: 'recurse'
+    merge_lists: 'false'
+    values:
+      pkg:
+        name: 'some-package'
+      config: '/path/to/a/configuration/file'
+    ...
+
+You can use `Jinja`_ as with any SLS files:
+
+.. code-block:: yaml
+
+    ---
+    strategy: 'overwrite'
+    merge_lists: 'true'
+    values:
+      output_dir: /tmp/{{ grains['id'] }}
+    ...
```

Comment:
Similarly:
```suggestion
.. code-block:: yaml

    ---
    strategy: 'recurse'
    merge_lists: 'false'
    values:
      TEMPLATE:
        pkg:
          name: 'some-package'
        config: '/path/to/a/configuration/file'
    ...

You can use `Jinja`_ as with any SLS files:

.. code-block:: yaml

    ---
    strategy: 'overwrite'
    merge_lists: 'true'
    values:
      TEMPLATE:
        output_dir: /tmp/{{ grains['id'] }}
    ...
```

Reply:
There is no `tplroot` under `values` in the YAML files.

Reply:
I can only go on my experiments but I have to use the `tplroot` in the yaml file to get my parameters to be effective:

https://gist.github.com/dafyddj/35dd3fc7e9bcef47112518c62f582bc7

Reply:
Erf, in fact it's a specificity of `openssh-formula` which [use subkeys for different components](https://github.com/saltstack-formulas/openssh-formula/pull/191/files#diff-26dfd111adaa5f111e2dd6df3c7c515cabd1eb96c7047695d2b8fc29efd16b6f).

In YAML files, the `values` key match the `mapdata` variable, so the following:

```yaml
values:
  openssh:
    foo: True
```

Is `mapdata.openssh.foo`.

Most formulas are like

```yaml
values:
  parameter: value
```

which became `mapdata.parameter`.

Hope it helps.

Reply:
Ok, thanks for the explanation. I guess it will all make more sense when we have it working on a few different formulas.